### PR TITLE
Provide option to restrict X-Forwarded-* headers to configured proxy addresses

### LINKED ISSUE #29888: Restricting X-Forwarded-* header parsing
### Discussed in https://github.com/quarkusio/quarkus/discussions/29887

<div type='discussions-op-text'>

<sup>Originally posted by **Doctor-love** December 15, 2022</sup>
Hi y'all,

Is it possible in Quarkus to restrict from which IP addresses the X-Forwarded-*  headers are parsed/trusted?
A similar option exist in nginx called ["set_real_ip_from"](https://nginx.org/en/docs/http/ngx_http_realip_module.html#set_real_ip_from).  
  
My use-case is exposing a distributed application both via a load balancer and direct connections, but for traceability I don't want to enable trivial spoofing of source IP addresses. Something like "_quarkus.http.proxy.allow-forwarded-from=ip.address.of.loadbalancer_" would be useful.</div>

### PR DESCRIPTION
**Pull request by michalvavrik**
closes: #29888

User can restrict proxy address from which Quarkus accepts `X-Forwarded` and `X-Forwarded-*` headers. It's possible to specify multiple proxy addresses as IPv4, IPv6, CIDR and hostnames. Hostnames must be resolved asynchronously to IP during the request, which makes it less performant option; we don't provide caching and advise users to use IP or CIDR.

### CODE REVIEW

**Review by sberyozkin — APPROVED** [review]
@michalvavrik Very good research, and a good PR. I've actually looked for two days in a row, but can not see what to suggest to improve. LGTM,
However I think we need at least one more review to see if something can be optimized in the way the checks are done for some proxy addresses. 
@stuartwdouglas, can you please review ?
Also CC @cescoffier  @pmlopes   
thanks

**Comment by Doctor-love** [comment]
Thanks for your work on this @michalvavrik , much appreciated!

Spent a bit of time testing this feature - took me quite a while to get started as I'm not a Java developer. I wrote a small program to return the requester's IP address:

```java
package org.example;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.inject.Inject;
import io.vertx.ext.web.RoutingContext;

@Path("/q")
public class RemoteAddressResource {

    @Inject
    RoutingContext context;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String returnRemoteAddress() {
        return context.request().remoteAddress().host();
    }
}
```

By all my test, it seems to work as described. I do however question the choice to only perform this validation for proxy headers of the "X-Forwarded-*" kind and not for others (such as "Forwarded") - is there are reason? Besides not being able to use this restriction for the standard "forwarded" header, a configuration such as this leaves the application exposed to spoofing ("X-Forwarded-For" is restricted but not "Forwarded"):

```
quarkus.http.proxy.proxy-address-forwarding=true
quarkus.http.proxy.allow-forwarded=true
quarkus.http.proxy.allow-x-forwarded=true
quarkus.http.proxy.trusted-x-forwarded-proxies=172.17.0.0/24
```

(I don't know why you would configure it this way and you get a warning, but seems like unnecessary room for error)

**Comment by michalvavrik** [comment]
Thank you @Doctor-love , your time is well appreciated!

re `Forwarded` - I simply implemented it only for `X-Forwarded` as that's what issue mentioned. This mechanism I provided can be refactored in order to support `Forwarded` too. IMO - this PR can't implement anything that is not described in the issue as changes need to be discussed and there should be consensus and there should be interest from users for the feature. Maybe we can do it later in a separate PR as I'm quite busy these days (provided there is opened issue).

re `quarkus.http.proxy.allow-forwarded=true` - as you said, you have been warned :-)

 I added comment to https://github.com/quarkusio/quarkus/issues/29888, let's move `Forwarded` discussion there.

**Comment by geoand** [comment]
What's the status of this one?

**Comment by michalvavrik** [comment]
We are waiting for @stuartwdouglas review.

**Comment by geoand** [comment]
As @stuartwdouglas is likely busy on other things, @pmlopes can you perhaps take a look please?

**Review by cescoffier — APPROVED** [review]
LGTM

**Comment by michalvavrik** [comment]
ping @sberyozkin

File: core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java
Comment (stuartwdouglas) [inline core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java:76]: I don't think this should be in core.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
Comment (stuartwdouglas) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:561]: This is complex enough that it should really be a top level class.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java
Comment (stuartwdouglas) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:131]: Shouldn't this check also be for the forwarded header?

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java
Comment (michalvavrik) [reply core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java:76]: I inspected config generation and to my knowledge this is the only way to do it currently.

Comment (michalvavrik) [reply core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java:76]: (btw. it's about config documentation generation - expected value TrustedXForwardedProxyCheckPart won't say too much..)

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java]: Done, thx.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:131]: Personally, I'd do it for forwarded only if there is interest from users.

**Comment by michalvavrik** [comment]
Failures are unrelated and fixed by this https://github.com/quarkusio/quarkus/pull/30509

**Comment by Ladicek** [comment]
I think you actually meant #30544

**Comment by michalvavrik** [comment]
Ah, sorry @Ladicek , I wrote the comment when I read your PR. Yes, I meant https://github.com/quarkusio/quarkus/pull/30544.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:131]: @michalvavrik It makes sense to have both types of headers treated the same way as far as this restriction is concerned, agree with Stuart; sorry for a delay

File: core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java
Comment (sberyozkin) [reply core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java:76]: @michalvavrik Sorry, what exactly you are trying to achieve with adding a ref to `io.quarkus.vertx.http.runtime.TrustedXForwardedProxyCheck.TrustedXForwardedProxyCheckPart` in the core annotations processor ? So that a more `typed` config property is generated ? May be it is simpler just to use one f the ConfigItem doc attributes ?

Comment (michalvavrik) [reply core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java:76]: Documentation table with configuration options contain property type. The property type in this case would be TrustedXForwardedProxyCheckPart and it would not provide users sufficient information on what should be added. There is no such `@ConfigItem` attribute to my knowledge. Constant above tells the processor how should it translate data type to user-friendly documentation that tells exactly what to do. For example, the option I added tells to use host, host[:port] or CIDR, which are all acceptable value formats.

If in doubt, please lookup property `quarkus.kafka-streams.bootstrap-servers` in generated documentation and be aware, that should `InetSocketAddress` not be there too, you would see `java.net.InetSocketAddress`.

@sberyozkin

Comment (sberyozkin) [reply core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java:76]: @michalvavrik Thanks, yes, it just seems like it is constant is added in the wrong place. It is a string but it indeed looks a bit strange to have a vertx package prefixed property. 
I'd not really worry about having just `TrustedXForwardedProxyCheckPart` generated, `ConfigItem` has an attribute allowing to have some extra docs attached, don't recall the name now, and there you can clarify what this type means. Does it sound reasonable ?

Comment (sberyozkin) [reply core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java:76]: Might also make sense to have TrustedXForwardedProxyCheckPart a standalone interface but it is up to you if you decide to keep it the current way

Comment (michalvavrik) [reply core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java]: I dropped the constant. Property description makes it clear what are expected values. Personally, I thought previous state with constant was user friendly, but let's proceed.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java]: Done.

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**
